third_party/move/move-compiler-v2/src/pipeline/dead_store_elimination.rs

```diff
@@ -1,57 +1,202 @@
+/// A node can represent both a definition and a use: such a node can only be of the
+/// form `Assign(dst, src)`, which follows from the above restrictions.
+/// As such, many code offsets in a function that do not meet the above criteria
+/// may not be present in this graph.
+///
+/// A side-effect-free definition can be removed safely if it is not alive later.
+struct DefUseGraph {
+    children: BTreeMap<CodeOffset, BTreeSet<CodeOffset>>,
+    parents: BTreeMap<CodeOffset, BTreeSet<CodeOffset>>,
+    dead: BTreeSet<CodeOffset>,
+}
+
+impl DefUseGraph {
+    /// Create a new `DefUseGraph` from the function `target`.
+    pub fn new(target: &FunctionTarget) -> Self {
+        let mut this = Self {
+            children: BTreeMap::new(),
+            parents: BTreeMap::new(),
+            dead: BTreeSet::new(),
+        };
+        this.populate_from(target);
+        this
+    }
+
+    /// Obtain the set of dead stores, i.e., code offsets which can be removed safely.
+    pub fn dead_stores(mut self) -> BTreeSet<CodeOffset> {
+        let mut dead = BTreeSet::new();
+        while let Some(offset) = self.remove_a_dead_node() {
+            dead.insert(offset);
+        }
+        dead
+    }
+
+    /// Populate an empty graph from the (restricted) definitions and uses in `target`.
+    fn populate_from(&mut self, target: &FunctionTarget) {
+        let code = target.get_bytecode();
+        let live_vars = target
+            .get_annotations()
+            .get::<LiveVarAnnotation>()
+            .expect("live variable annotation is a prerequisite");
+        for (offset, instr) in code.iter().enumerate() {
+            use Bytecode::*;
+            match instr {
+                Assign(_, dst, src, _) if dst == src => {
+                    // self-assignment is always a dead store
+                    self.incorporate_definition(*dst, offset as CodeOffset, live_vars, true);
+                },
+                Assign(_, dst, ..) | Load(_, dst, _) => {
+                    self.incorporate_definition(*dst, offset as CodeOffset, live_vars, false);
+                },
+                _ => {},
+            }
+        }
+    }
+
+    /// Remove one dead node (arbitrary but deterministic) from the graph, if present.
+    fn remove_a_dead_node(&mut self) -> Option<CodeOffset> {
+        if let Some(node) = self.dead.pop_last() {
+            let parents = self.disconnect_from_parents(node);
+            let children = self.disconnect_from_children(node);
+            // Reconnect all parents to all children.
+            for parent in parents.iter() {
+                for child in children.iter() {
+                    self.children.entry(*parent).or_default().insert(*child);
+                    self.parents.entry(*child).or_default().insert(*parent);
+                }
+            }
+            // Parents are the only ones who could become dead (if `node` was their last child).
```

Comment:
So we have no way to recognize a cycle of dead assignments?

Reply:
Tracking it here: https://github.com/aptos-labs/aptos-core/issues/12400

The algo we have can be extended to do so, but will do in a follow up PR. If there is a cycle of dead assignments with none of them having usages outside of the cycle, then I think we can remove it. There are some related cases where you can go further, so will explore that too.